Support artifact formats in `contract!` macro

taminomara opened the pull request:
Part of #512, merge after #555.

Refactor `contract!` macro a bit, add support for artifact formats, rework contract renaming syntax (see discussion in #512).

Commit:
Support hardhat artifacts in macro

taminomara commented:
Working on test cases now.

Commit:
Tests

Commit:
Rustfmt

taminomara commented:
Tests failed with a linker error. I'm not sure if it's a transient error, and how to fix it.

```
  = note: /usr/bin/ld: /home/runner/work/ethcontract-rs/ethcontract-rs/target/debug/deps/ethcontract_derive-a6524bc4a7549b69.ethcontract_derive.bi7dxuh2-cgu.4.rcgu.o: in function `<T as alloc::string::ToString>::to_string':
          /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/alloc/src/string.rs:2277: undefined reference to `core::result::Result<T,E>::expect'
          /usr/bin/ld: /home/runner/work/ethcontract-rs/ethcontract-rs/target/debug/deps/ethcontract_derive-a6524bc4a7549b69.ethcontract_derive.bi7dxuh2-cgu.7.rcgu.o: in function `<T as alloc::string::ToString>::to_string':
          /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/alloc/src/string.rs:2277: undefined reference to `core::result::Result<T,E>::expect'
```

nlordell commented on ethcontract-derive/src/lib.rs:
```diff
@@ -26,69 +27,166 @@ use syn::{
+/// contract!("https://my.domain.local/path/to/contract.json")
+///
+/// // etherscan.io
+/// contract!("etherscan:0xC02AAA39B223FE8D0A0E5C4F27EAD9083C756CC2");
```
Ubernit: We can use a checksumed address :stuck_out_tongue: 
```suggestion
/// contract!("etherscan:0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2");
```

nlordell commented on ethcontract-derive/src/lib.rs:
````diff
@@ -26,69 +27,166 @@ use syn::{
 /// ```
 ///
-/// Note that Etherscan rate-limits requests to their API, to avoid this an
+/// Note that etherscan rate-limits requests to their API, to avoid this an
````
Shouldn't Etherscan be capitalized as it is a proper noun?

taminomara replied:
I've done this to match capitalisation with other parts of code. I can change it separately.

nlordell reviewed (approved):
Looks good.

nlordell commented:
> Tests failed with a linker error

Maybe try and clear the workflow cache and see if that helps? See https://github.com/gnosis/ethcontract-rs/pull/521.

vkgnosis commented:
> Tests failed with a linker error.

Does `cargo test` work locally? Either way it seems like a compiler bug. Maybe related to the cache that is used in CI. If it persists we can try clearing the cache.

taminomara commented:
Tests do pass locally. I'll re-run tests and see if we need to reset caches.

Commit:
Update ethcontract-derive/src/lib.rs

Co-authored-by: Nicholas Rodrigues Lordello <nlordell@gmail.com>

nlordell commented on ethcontract-derive/src/lib.rs:
````diff
@@ -26,69 +27,166 @@ use syn::{
+///
+///   ```ignore
+///   contract!(
+///       "etherscan:0xC02AAA39B223FE8D0A0E5C4F27EAD9083C756CC2",
````
Here too :stuck_out_tongue: 
```suggestion
///       "etherscan:0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2",
```

vkgnosis commented:
It still fails. To clean the cache do the same thing as in this commit https://github.com/gnosis/ethcontract-rs/pull/521/commits/7957678134d2e8c29f898914ab260e6b5d188055 .

Commit:
One more review suggestion

taminomara commented:
It still fails for some reason...

vkgnosis commented:
Internet searching around a bit it is similar to some issues in the rustc repository so I'm confident this is a compiler bug. Probably triggered by some of the details of the image that is used by github actions. Not sure how to fix it. We could try changing some stuff around like different image, changing compiler version. Another random idea is swapping the cargo test and cargo clippy lines here https://github.com/gnosis/ethcontract-rs/blob/main/.github/workflows/pull-request.yaml#L41 .

taminomara commented:
All other PRs fail with the same error, so that's definitely not related to changes from this one.

taminomara commented:
I'll create a new PR and try different things.

taminomara commented:
So, I've tried changing things in workflow, including new compiler version, with no success. I've also tried to re-run some old jobs that were green, and they have also failed. It's definitely a bug somewhere. What's strange is that I don't see any changes since last week: the same compiler version, the same OS, the same image version, everything.

vkgnosis commented:
Very weird. Maybe Github changed something in the image? It says "Ubuntu 20.04.2  LTS" but they might have changed some installed packages. I'll try to play around with it in a local container and see if I can reproduce or fix the issue.

vkgnosis commented:
I am able to reproduce this locally. I started with that same Ubuntu docker image (but not with the packages github installs) and then added everything needed for Rust.
The one workaround I found is instead of running `cargo test` at the top level we can run it in every workspace folder separately. Not perfect but at least it fixes CI.
I'll see if i can minimize it and report the bug to rustc or cargo.
